File: web/blocks/tags/inherit-tags.dart
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

part of NetTango;

class InheritTags extends AllowedTags {
  Clause clause;

  InheritTags(Clause clause) {
    this.clause = clause;
  }

  AllowedTags clone(Clause newClause) {
    return new InheritTags(newClause);
  }

  bool check(Iterable<Block> blocks) {
    final parent = InheritTags.getConcreteTags(this.clause);
    if (parent == null) {
      // no restrictions at the moment
      return true;
    }
    return parent.check(blocks);
  }

  static ConcreteTags getConcreteTags(Clause clause) {
    switch (clause.owner.dragData.parentType) {

      case "workspace-chain":
        final first = clause.owner.workspace.chains[clause.owner.dragData.chainIndex].blocks.first;
        return first.canBeStarter ? first.allowedTags : null;

      case "block-clause":
        final ownerClause = clause.owner.workspace.getBlockInstance(clause.owner.dragData.parentInstanceId).clauses[clause.owner.dragData.clauseIndex];
        if (ownerClause.allowedTags is InheritTags) {
          return InheritTags.getConcreteTags(ownerClause);
        }
        return ownerClause.allowedTags;

      // I guess a new block is a fragment?  But we shouldn't be calling this, really.
      case "new-block":
        return null;

      default:
        throw new Exception("Unknown block parent type: ${clause.owner.dragData.parentType}");

    }
  }

}

File: src/blocks/tags/inherit-tags.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import { ConcreteTags } from "../../types/types"
import { Block } from "../block"
import { Clause } from "../clause"
import { ChainDragData } from "../drag-drop/drag-data/chain-drag-data"
import { ClauseDragData } from "../drag-drop/drag-data/clause-drag-data"
import { NewDragData } from "../drag-drop/drag-data/new-drag-data"
import { checkConcreteTags } from "./concrete-tags"

function getConcreteTags(clause: Clause): ConcreteTags | null {
  if (clause.owner.dragData === null) {
    throw new Error("No drag data to use for tags")
  }

  const data = clause.owner.dragData
  if (data instanceof ChainDragData) {
    const first = clause.owner.workspace.chains[data.chainIndex].blocks[0]
    return first.canBeStarter ? (first.b.allowedTags as ConcreteTags) : null
  }

  if (data instanceof ClauseDragData) {
    const ownerClause = clause.owner.workspace.getBlockInstance(data.parentInstanceId).clauses[data.clauseIndex]
    if (ownerClause.c.allowedTags.type === 'inherit') {
      return getConcreteTags(ownerClause)
    }
    return (ownerClause.c.allowedTags as ConcreteTags)
  }

  // I guess a new block is a fragment?  But we shouldn't be calling this, really.
  if (data instanceof NewDragData) {
    return null
  }

  throw new Error(`Unknown block drag data type: ${clause.owner.dragData}`)
}

function checkInheritTags(clause: Clause, blocks: Block[]): boolean {
  const parent = getConcreteTags(clause)
  if (parent === null) {
    // no restrictions at the moment
    return true
  }
  return checkConcreteTags(parent, blocks)
}

export { checkInheritTags }
